tools/codegen/dest/gen_external_aten_fallbacks.py

```diff
@@ -0,0 +1,305 @@
+from tools.codegen.model import (Argument, ExternalBackendFunction,
+                                 ExternalBackendFunctionsGroup, SchemaKind,
+                                 assert_never, Return, is_generic_dispatch_key,
+                                 ListType, OptionalType, BaseType, BaseTy, Variant)
+from tools.codegen.api.types import DispatcherSignature, CppSignatureGroup
+import tools.codegen.api.dispatcher as dispatcher
+import tools.codegen.api.cpp as cpp
+
+# TODO: this contains a list of regex for ops that don't get a CPU fallback.
+# We should just register fallthroughs when we make the CPU fallback a boxed kernel.
+_FN_DENYLIST_REGEX = [
+    # ATEN functions
+    r'[^(]*cudnn',
+    r'slow_conv_transpose2d_backward.grad_output',
+    r'slow_conv_transpose3d_backward.grad_output',
+    r'slow_conv3d_backward.grad_input',
+    r'thnn_conv2d_backward.grad_input',
+    r'thnn_conv_depthwise2d_backward.grad_input',
+    # XLA/TPU functions
+]
+
+# TODO: remove this list.
+# Instead, the codegen will figure out which ops to generate _out wrappers for
+# entirely from the yaml. Maintaining the same behavior as current XLA codegen for now.
+_FN_OUT = [
+    'abs',
+    'add',
+    'acos',
+    'acosh',
+    'asin',
+    'asinh',
+    'atan',
+    'atan2',
+    'atanh',
+    'baddbmm',
+    'bernoulli',
+    'binary_cross_entropy',
+    'binary_cross_entropy_backward',
+    'clamp',
+    'div',
+    'gather',
+    'ger',
+    'hardsigmoid',
+    'kthvalue',
+    'index_select',
+    'inverse',
+    'log',
+    'masked_select',
+    'maximum',
+    'minimum',
+    'pow',
+    'prod',
+    'nonzero',
+    'round',
+    'normal',
+    'std',
+    'take',
+    'topk',
+    'var',
+]
+
+def requires_backend_wrapper(f: ExternalBackendFunction) -> bool:
+    requires_lowering = not any(is_generic_dispatch_key(k) for k in f.native_function.dispatch)
+    has_xla_lowering = f.metadata is not None
+    in_denylist = any([re.match(frx, str(f.native_function.func.name)) for frx in _FN_DENYLIST_REGEX])
+    return not in_denylist and (requires_lowering or has_xla_lowering)
+
+def xla_tensor_creation_api(ret_name: str, ret: Return, device_param_name: str, *, tuple_idx: Optional[int] = None) -> str:
+    if ret.type == BaseType(BaseTy.Tensor) and not ret.is_write:
+        # Only raw Tensor (non-reference) returns need to go through the XLA tensor creation API.
+        # Tensor references can be returned directly, since they've already been converted to XLA tensors.
+        # See Note [Tensor Copy Returns]
+        bridge_api = 'CreateXlaTensor'
+    elif isinstance(ret.type, ListType) and ret.type.elem == BaseType(BaseTy.Tensor):
+        bridge_api = 'CreateXlaTensors'
+    else:
+        # for non tensor-types, there's no need to wrap the output in an xla bridge api.
+        return ret_name
+
+    out_name = 'x_result'
+    if tuple_idx is not None:
+        out_name = f"std::get<{tuple_idx}>(x_result)"
```

Comment:
The `x_result` embedded here should probably be `{out_name}`... more importantly though, you probably want to hoist `x_result` up out of the various string literals it's embedded in and turn it into a global, since it's a shared point of definition for the codegen.

Reply:
agreed.

Actually, I don't think it needs to be a global; it's used by both the cpu fallback and the codegen'd `out` wrappers, but there's no reason that the variable name needs to be the same between the two aside from byte-for-byte compatibility. The important thing is just that it's defined in one place / consistent everywhere within the CPU fallback, and everywhere within the `out` wrapper.

So lmk if you have a different opinion, but I'm leaning towards defining it twice, once in the `out`-wrapper codegen and once in the CPU fallback codegen. That'll also make things less messy later, when I move the `out` wrapper codegen into a different file from the fallback.

Reply:
oh scratch that, I just double checked and `x_result` isn't actually needed in the `out` wrapper codegen. It's only used in the CPU fallback.